libbpf-tools/memleak.c

```diff
@@ -0,0 +1,707 @@
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	return vfprintf(stderr, format, args);
+}
+
+static int get_libc_path(char *path)
+{
+	FILE *f;
+	char buf[PATH_MAX] = {};
+	char *filename;
+	float version;
+
+	f = fopen("/proc/self/maps", "r");
+	if (!f)
+		return -errno;
+
+	while (fscanf(f, "%*x-%*x %*s %*s %*s %*s %[^\n]\n", buf) != EOF) {
+		if (strchr(buf, '/') != buf)
+			continue;
+		filename = strrchr(buf, '/') + 1;
+		if (sscanf(filename, "libc-%f.so", &version) == 1) {
+			memcpy(path, buf, strlen(buf));
+			fclose(f);
+			return 0;
+		}
+	}
+
+	fclose(f);
+	return -1;
+}
+
+struct alloc_stack {
+	uint64_t stack_id;
+	int size;
+	int nr;
+};
+
+static int partition(struct alloc_stack *list, int low, int high)
+{
+	struct alloc_stack pivot = list[high];
+	struct alloc_stack temp;
+	int i = low - 1;
+	int j;
+
+	for (j = low; j <= high - 1; j++) {
+		if (list[j].size > pivot.size) {
+			i++;
+			temp = list[j];
+			list[j] = list[i];
+			list[i] = temp;
+		}
+	}
+
+	temp = list[i + 1];
+	list[i + 1] = list[high];
+	list[high] = temp;
+	return (i + 1);
+}
+
+static void quick_sort(struct alloc_stack *list, int low, int high)
+{
+	int pi;
+
+	if (list == NULL) return;
+
+	if (low < high) {
+		pi = partition(list, low, high);
+		quick_sort(list, low, pi - 1);
+		quick_sort(list, pi + 1, high);
+	}
+}
+
+static void print_combined_outstanding_allocations(struct ksyms *ksyms, struct syms_cache *syms_cache,
+		struct memleak_bpf *obj)
+{
+	uint64_t stack_id = 0, next_stack_id;
+	const struct ksym *ksym;
+	const struct syms *syms;
+	const struct sym *sym;
+	int err, i, ifd, sfd, cnt;
+	unsigned long *ip;
+	time_t now;
+	struct tm * timeinfo;
+	struct combined_alloc_info_t info;
+	struct alloc_stack *stacks;
+	int max_prints = env.top;
+
+	time (&now);
+	timeinfo = localtime(&now);
+	printf("[%02x:%02d:%02d] Top %d stacks with outstanding allocations:\n",
+		timeinfo->tm_hour, timeinfo->tm_min, timeinfo->tm_sec, env.top);
+
+	ip = calloc(env.max_stack_depth, sizeof(*ip));
+	if (!ip) {
+		fprintf(stderr, "failed to alloccate memory of stack trace\n");
+		return;
+	}
+
+	stacks = calloc(MAX_HASH_ENTRY_NUM, sizeof(*stacks));
+	if (stacks == NULL) {
+		fprintf(stderr, "failed to allocate memory for stack info\n");
+		goto cleanup;
+	}
+	memset(stacks, 0, MAX_HASH_ENTRY_NUM * sizeof(*stacks));
+
+	cnt = 0;
+	ifd = bpf_map__fd(obj->maps.combined_allocs);
+	sfd = bpf_map__fd(obj->maps.stack_traces);
+	while (!bpf_map_get_next_key(ifd, &stack_id, &next_stack_id)) {
+		err = bpf_map_lookup_elem(ifd, &next_stack_id, &info);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup info: %d\n", err);
+			goto cleanup;
+		}
+		stack_id = next_stack_id;
+
+		stacks[cnt].stack_id = next_stack_id;
+		stacks[cnt].size = info.total_size;
+		stacks[cnt].nr = info.number_of_allocs;
+		cnt++;
+	}
+
+	quick_sort(stacks, 0, cnt);
+	if (cnt < max_prints) {
+		max_prints = cnt;
+	}
+
+	for (i=0; i < max_prints; i++) {
```

Comment:
`i = 0`.

Reply:
Fix in commit [2beb289e](https://github.com/iovisor/bcc/pull/3828/commits/2beb289e98d651bddc2127299c5627b5d6149fce).